### ISSUE TITLE
How to make Hanami::Entity behave like Dry::Struct?

### ISSUE THREAD

**Issue by korolvs** [hanami/model#411]
For example, I have classes A and B:
```
class A < Dry::Struct
   attribute :id, Types::Strict:Int
end

class B < Hanami::Entity
  attributes do
     attribute :id, Types::Strict:Int
  end
end
```
If I create an empty instance of class A there will be an error:
```
> A.new
Dry::Struct::Error: [A.new] :id is missing in Hash input
```
But if I create an empty instance of class B - no:
```
> B.new
=> #<B:0x007f95105b8658 @attributes={}>
```

How can I get the same error for class B?
Thanks

**Comment by cllns** [comment]
`Dry::Struct` (through `Dry::Types`) is more flexible than `Hanami::Entity`. `Dry::Types` let you specify this behavior, but `Hanami::Entity` doesn't. See the `constructor_type` option for `Dry::Types` http://dry-rb.org/gems/dry-types/hash-schemas/

If you need this behavior, can you use `dry-struct` for your Entities? I'm not 100% sure, but I think it should work. The API for `Hanami::Entity` is defining `id`, which you're already doing.

There's also a protected `attribute?` method, which is used by `method_missing` ([It also has `method_missing` defined](https://github.com/hanami/model/blob/master/lib/hanami/entity.rb#L142)). You might be able to avoid implementing those, I'm not sure.

```
> (A.new(id: 1).methods - Kernel.methods).sort 
=> [:[], :id, :to_h, :to_hash] # Dry::Struct methods

> (B.new.methods - Kernel.methods).sort 
=> [:attribute?, :id, :method_missing, :to_h, :to_hash] # Hanami::Entity methods
```

**Comment by korolvs** [comment]
@cllns Thank you for your answer!
I found that it is possible to use ```Dry::Struct``` for Entities like this:
```
class Base < Dry::Struct
  module Types
    include Hanami::Model::Types
  end

  # Dummy method
  def self.schema=(value)
  end
end

class A < Base
  attribute :id, Types::Int
  attribute :name, Types::String
end
```

**Comment by davydovanton** [comment]
@hanami/core should we add information about "dry-struct as an entity" to guides?

**Comment by jodosha** [comment]
@davydovanton Nope, I think we can do better.

The solution found by @korolvs is a hack, but that is a clear sign that we can make our entities more flexible by taking advantage of the new features from `dry-types`.

```ruby
class Book < Hanami::Entity
  attributes :strict do
    attribute :id, Types::Strict::Int
  end
end
```

**Comment by cllns** [comment]
Ooh, I like it @jodosha!

**Comment by jodosha** [comment]
@korolvs Would you like to give a try to #431?

**Comment by jodosha** [comment]
Implemented by #431
